src/transformers/integrations/hqq.py

```diff
@@ -0,0 +1,108 @@
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"HQQ (Half-Quadratic Quantization) integration file"
+
+import torch
+
+from ..utils import is_hqq_available, logging
+from ..utils.hqq_utils import autoname_modules, get_linear_tags, name_to_linear_tag
+
+
+if is_hqq_available():
+    from hqq.core.quantize import HQQLinear
+else:
+    HQQLinear = None
+
+logger = logging.get_logger(__name__)
+
+
+def _prepare_for_hqq_linear(model, patch_params, has_been_replaced, current_key_name=None):
+    if not is_hqq_available():
+        raise ValueError(
+            "HQQ is not available. Please follow the instructions to install it: `https://github.com/mobiusml/hqq/`"
+        )
+
+    for name, module in model.named_children():
+        if current_key_name is None:
+            current_key_name = []
+        current_key_name.append(name)
+
+        if isinstance(module, torch.nn.Linear):
+            # Get linear tag
+            linear_tag = name_to_linear_tag(module.name)
+
+            # We put the module quant_config into the nn.Linear layer so we can access it later in quantizer_hqq.create_quantized_param()
+            if linear_tag in patch_params:
+                if patch_params[linear_tag] is not None:
+                    model._modules[name].quant_config = patch_params[linear_tag]
+                    # Store the module class in case we need to transpose the weight later
+                    model._modules[name].source_cls = type(module)
+                    # Force requires grad to False to avoid unexpected errors
+                    model._modules[name].requires_grad_(False)
+
+            has_been_replaced = True
+
+        if len(list(module.children())) > 0:
+            _, has_been_replaced = _prepare_for_hqq_linear(
+                module,
+                patch_params=patch_params,
+                has_been_replaced=has_been_replaced,
+            )
+        # Remove the last key for recursion
+        current_key_name.pop(-1)
+
+    return model, has_been_replaced
+
+
+def prepare_for_hqq_linear(model, quantization_config=None, modules_to_not_convert=None, has_been_replaced=False):
+    """
+    Prepares nn.Linear layers for HQQ quantization.
+    Since each layer type can have separate quantization parameters, we need to do the following:
+    1- tag each module with its neme via autoname_modules()
+    2- Extract linear_tags (e.g. ['self_attn.q_proj', ...])
+    3- Map quantization parameters as a dictionary linear_tag -> quant_params as HQQLinear exepects it, this is referred to as patch_params
+    """
+
+    modules_to_not_convert = [] if modules_to_not_convert is None else modules_to_not_convert
+
+    # Add name to module
+    autoname_modules(model)
```

Comment:
It's a lot of stuff, there are like 8 methods, and some methods are used in other files. 
Should I still move the util functions inside hqq.py and for the other files import them from hqq.py?

Reply:
I think it's fine to move all of them here, for new quant methods we prefer to have on single file where all util methods related to that quant method live 🙏

Reply:
Do you think it's good to move them? see my comment above

Reply:
Ok so I moved the functions inside. The reason why I thought we needed is because some functions were re-used in the save/load logic But since we decided to not support that, they can be removed. I deleted the whole hqq_utils.py file.